ENT-11175: Rewrote the intro to be more clear about different types of projects (#178)
cfengine/cfbs: JSON.md

# CFEngine Build JSON file format

The `cfbs` command line tool uses `cfbs.json` configuration files.
We call the folder / git repo which has a `cfbs.json` file (and optionally other files and folders) a _project_, and the end result after combining modules in a project into policy a _policy set_.
The _policy set_ is what you deploy to your CFEngine hub / policy server, in `/var/cfengine/masterfiles`.

See the [README](https://github.com/cfengine/cfbs/blob/master/README.md) for general documentation on the `cfbs` tool and its commands.
This file is specifically about the `cfbs.json` file format, it should serve as detailed and useful information for more advanced users who are making their own modules or contributing to CFEngine Build.

> There are 3 types of `cfbs.json` files:

Reviewer: ```suggestion
The type of a project is specified in a top-level `type` key in `cfbs.json`.
There are 3 types of projects:
```

> * For building a policy set you create a project with type `policy-set`.

Reviewer: ```suggestion
* For building a policy set you create a project with a top-level key `type` with the value `policy-set`.
```

  This is the default when running `cfbs init`, and what most users encounter when first using the tool and CFEngine.

>   You then need to use the `build` key to specify which modules to use in `cfbs build`.

Reviewer: ```suggestion
  You then need to use the top-level `build` key to specify which modules to use in `cfbs build`.
```

Author: Hmm, to me it seems to be a bit more tedious / harder to read if that is repeated every time. Probably I'll try to accept some, but not all, of the suggestions.

> * The index of all available modules for `cfbs add <module-name>` has type `index`.

Reviewer: ```suggestion
* For specifying all available modules for `cfbs add <module-name>` add a top-level key `type` with value `index`.
```

  The available modules must be in a dictionary in the `index` field.
  By default, [this index available GitHub](https://github.com/cfengine/build-index/blob/master/cfbs.json) is used.

> * When developing your own reusable modules to use in other projects, `module` type is used.

Reviewer: ```suggestion
* When developing your own reusable modules to use in other projects, a top-level `type` with value `module` is used.
```

When `cfbs` is using the default index and when we build the [build.cfengine.com](https://build.cfengine.com) website, we use a separate [`versions.json`](https://github.com/cfengine/build-index/blob/master/versions.json) file to keep track of all the versions of modules, their tarballs and checksums.

> When contributors edit the index ([like this](https://github.com/cfengine/build-index/pull/465/files)), an automated PR is generated to make the appropriate edit to `versions.json` ([like this](https://github.com/cfengine/build-index/pull/466/files)), (after downloading and uploading the module), so users don't have to `versions.json` manually.

Reviewer: ```suggestion
When contributors edit the index ([like this](https://github.com/cfengine/build-index/pull/465/files)), an automated PR is generated to make the appropriate edit to `versions.json` ([like this](https://github.com/cfengine/build-index/pull/466/files)), (after downloading and uploading the module), so users don't have to update `versions.json` manually.
```

# CFEngine Build JSON file format

The `cfbs` command line tool uses `cfbs.json` configuration files.
We call the folder / git repo which has a `cfbs.json` file (and optionally other files and folders) a _project_, and the end result after combining modules in a project into policy a _policy set_.
The _policy set_ is what you deploy to your CFEngine hub / policy server, in `/var/cfengine/masterfiles`.

See the [README](https://github.com/cfengine/cfbs/blob/master/README.md) for general documentation on the `cfbs` tool and its commands.
This file is specifically about the `cfbs.json` file format, it should serve as detailed and useful information for more advanced users who are making their own modules or contributing to CFEngine Build.

The type of a project is specified in a top-level `type` key in `cfbs.json`.
There are 3 types of projects:

> * For building a policy set you create a project with type `policy-set`.

Reviewer 2: I would suggest enumarting these, starting with the key name. It means I don't have to read the whole sentence to look up which key we are talking about, and it is thus easier to use as a lookup table of sorts

Author: To me that seems more appropriate for the keys section below. In the intro I prefer just sentences.

Author: https://github.com/cfengine/cfbs/pull/180

  This is the default when running `cfbs init`, and what most users encounter when first using the tool and CFEngine.
  You then need to use the top-level `build` key to specify which modules to use in `cfbs build`.
* The index of all available modules for `cfbs add <module-name>` has type `index`.
  The available modules must be in a dictionary in the `index` field.

>   By default, [this index available GitHub](https://github.com/cfengine/build-index/blob/master/cfbs.json) is used.

Reviewer 3: See https://github.com/cfengine/cfbs/pull/179/files
